### PR TITLE
Issue 0041 - Add elastic wave propagation 

### PR DESCRIPTION
**Pull request by SouzaEM**
I'm creating a draft pull request for discussing the implementations/modifications and avoiding a big review. Initially I just added the directory structure and the input parameters for the elastic model.

I think we need a ElasticWave base class with common code for IsotropicElasticWave, TTLElasticWave, ViscoElasticWave, etc. The parsing of the parameters dictionary is used as an example to illustrate why I think the base class would help us adding new features.

### CODE REVIEW

File: spyro/io/model_parameters.py
Comment (SouzaEM) [inline spyro/io/model_parameters.py:1]: I removed some extra empty spaces from some warning messages. This is not related to the actual development of elastic waves.

File: spyro/solvers/wave.py
Comment (SouzaEM) [inline spyro/solvers/wave.py:1]: I renamed the `_get_initial_velocity_model` to `_initialize_model_parameters` because the elastic wave propagators have more parameters. I also moved the implementation to AcousticWave to allow ElasticWave to inherit Wave behavior consistently.

File: spyro/solvers/acoustic_wave.py [hunk 1]
```diff
@@ -144,3 +144,33 @@ def reset_pressure(self):
...
+            vp_filename, vp_filetype = os.path.splitext(
+                self.initial_velocity_model_file
+            )
+            warnings.warn("Converting segy file to hdf5")
+            write_velocity_model(
+                self.initial_velocity_model_file, ofname=vp_filename
+            )
+            self.initial_velocity_model_file = vp_filename + ".hdf5"
+
+        if self.initial_velocity_model_file.endswith((".hdf5", ".h5")):
```
Comment (SouzaEM) [inline spyro/solvers/acoustic_wave.py:168]: I added `.h5` as an alternative to `.hdf5` because it is another common extension for HDF5.

File: spyro/utils/typing.py
Comment (Olender) [inline spyro/utils/typing.py:1]: This can be done as soon as Firedrake fixes the GLL quads numbering issue, since it is compatible with 3.12. If they take too long I can look into it after turning in my thesis.

File: spyro/solvers/wave.py [hunk 2]
```diff
@@ -223,35 +223,10 @@ def _map_sources_and_receivers(self):
...
-            )
-
-        if self.debug_output:
-            fire.File("initial_velocity_model.pvd").write(
-                self.initial_velocity_model, name="velocity"
-            )
+    
+    @abstractmethod
+    def _initialize_model_parameters(self):
+        pass
```
Comment (Olender) [inline spyro/solvers/wave.py:231]: Good idea moving this to the other classes

File: spyro/solvers/wave.py [hunk 3]
```diff
@@ -315,3 +290,51 @@ def set_last_solve_as_real_shot_record(self):
...
+    def _get_prev_vstate(self):
+        pass
+
+    @abstractmethod
+    def _set_next_vstate(self, vstate):
+        pass
+
+    @abstractmethod
+    def _get_next_vstate(self):
+        pass
```
Comment (Olender) [inline spyro/solvers/wave.py:318]: I really like what you have done here. Your changes significantly improve the readability and maintainability of the subsequent time integration methods. However, to ensure there are no unintended increases in memory usage or runtime, I recommend we generate flame graphs for verification.

Comment (SouzaEM) [reply spyro/solvers/wave.py:318]: I believe it would not increase the runtime considerably because the numerical parts of the code are probably the bottlenecks. I will work on the flame graphs to make sure that these modifications did not degrade the code performance.

File: spyro/utils/typing.py
Comment (SouzaEM) [reply spyro/utils/typing.py:1]: This is not a really important feature (we don't need to set it as a priority), but I think it helps following up which methods are overriding something from the parent class.

File: spyro/solvers/wave.py (hunk 2 above)
Comment (SouzaEM) [reply spyro/solvers/wave.py:231]: I think it will help because other models have more parameters (e.g., density and S-wave velocity).

File: spyro/solvers/wave.py (hunk 3 above)
Comment (Olender) [reply spyro/solvers/wave.py:318]: It's been a while since I last did a flame graph, but I found this resource from the Firedrake project to be quite useful for code optimization: https://www.firedrakeproject.org/optimising.html. You don't have to do it, you can just time the 3d test and the time_convergence test locally to see if nothing major changed. When we go to elastic 3D, we should generate flamegraphs to try and find any unnecessary bottlenecks

File: spyro/solvers/acoustic_solver_construction_no_pml.py
Comment (Olender) [inline spyro/solvers/acoustic_solver_construction_no_pml.py:1]: Great!

File: spyro/solvers/mms_acoustic.py [hunk 4]
```diff
@@ -66,3 +69,7 @@ def analytical_solution(self, t):
         # self.analytical.assign(analytical)
 
         return self.analytical
+    
+    @override
+    def update_source_expression(self, t):
+        self.q_t.assign(2*t)
```
Comment (Olender) [inline spyro/solvers/mms_acoustic.py:75]: I'm uncertain if this is the best practice. While using MMS for this small problem might not cause any issues, for larger cases in the future, we should prioritize wrapping scalar time-dependent variables inside a `fire.Constant` for better performance.

File: spyro/solvers/time_integration_central_difference.py [hunk 5]
```diff
@@ -44,11 +44,16 @@ def central_difference(wave, source_id=0):
     usol_recv = []
     save_step = 0
     for step in range(nt):
-        rhs_forcing.assign(0.0)
+        # Basic way of applying sources
+        wave.update_source_expression(t)
```
Comment (Olender) [inline spyro/solvers/time_integration_central_difference.py:48]: This method needs to just pass when MMS isn't used. Assigning something takes a while, it is really significant in 3D.

Comment (Olender) [reply spyro/solvers/time_integration_central_difference.py:48]: Nevermind, I saw that it just passes when MMS isn't used in the code down below

File: spyro/solvers/mms_acoustic.py (hunk 4 above)
Comment (SouzaEM) [reply spyro/solvers/mms_acoustic.py:75]: I didn't get the issue here. The variable "q_t" is defined as a `Constant` is line 14. This should be faster than before because the method was returning new constants in every call. As far as I remember, using the same constant avoids some parts of JIT or assembling.

Comment (Olender) [reply spyro/solvers/mms_acoustic.py:75]: Ah, I see now. I mistakenly thought that `q_t` was defined as a Function rather than a Constant, I didn't see line 14. Thanks for clarifying. Yes, reusing the same constant does indeed optimize the process by avoiding the overhead associated with redefining it repeatedly, leading to better performance.

**Review by Olender — APPROVED** [review]
